# SNPs_Analysis_usingR
This R script is used to translate the nucleotide changes generated using NUCMER. The original script is written by Federico M. Giorgi (Associate Professor Department of Pharmacy and Biotechnology, University of Bologna Bologna, Italy). The script was modified by Ankeet Kumar (PhD Candidate at the Department of Biochemistry, Indian Insitute of Science, 560012, Bengaluru-KA).

The Nucmer tool is used to generate SNPs based on the pairwise comparison of genome sequences to a reference sequence. The outcome of nucmer is a table containing SNPs with the position of nucleotide change and nucleotide variation. We have developed various versions of the script which can handle genomes with genes in only positive polarity, overlapping genes (nipah virus) and also a script which has capability to analyse ambipolar genomes with overlapping sequences (mpox virus)

The script has been used in the following publications:
1. Yadav, P., Kumar, A., Nath, S.S. et al. Unravelling the genomic origins of lumpy skin disease virus in recent outbreaks. BMC Genomics 25, 196 (2024). https://doi.org/10.1186/s12864-024-10061-3
2. A. Kumar, M. Kaustubh, S.S. Nath, U. Tatu, Identification of clade-specific single nucleotide polymorphisms for improved Rabies virus surveillance in Canis familiaris host, BioRxiv (2023) 2023.08.25.553658. https://doi.org/10.1101/2023.08.25.553658.
3. Kumar A, Jhanwar P, B R, Gulati A, Tatu U. Genomic analyses of recently emerging clades of mpox virus reveal gene deletion and single nucleotide polymorphisms that correlate with altered virulence and transmission. BioRxiv 2024:2024.09.24.614696. https://doi.org/10.1101/2024.09.24.614696.
